pkg/core/native/native_test/cryptolib_verification_test.go

```diff
@@ -0,0 +1,924 @@
+	// Pack public keys and store at LOC1.
+	for _, pub := range pubs {
+		emit.Bytes(vrf.BinWriter, pub.Bytes())
+	}
+	emit.Int(vrf.BinWriter, int64(len(pubs)))
+	emit.Opcodes(vrf.BinWriter, opcode.PACK, opcode.STLOC1)
+
+	// Get message and store it at LOC2.
+	// msg = [4-network-magic-bytes-LE, tx-hash-BE]
+	emit.Syscall(vrf.BinWriter, interopnames.SystemRuntimeGetNetwork) // push network magic (Integer stackitem), can have 0-5 bytes length serialized.
+	// Convert network magic to 4-bytes-length LE byte array representation.
+	emit.Int(vrf.BinWriter, 0x100000000)
+	emit.Opcodes(vrf.BinWriter, opcode.ADD, // some new number that is 5 bytes at least when serialized, but first 4 bytes are intact network value (LE).
+		opcode.PUSH4, opcode.LEFT) // cut the first 4 bytes out of a number that is at least 5 bytes long, the result is 4-bytes-length LE network representation.
+	// Retrieve executing transaction hash.
+	emit.Syscall(vrf.BinWriter, interopnames.SystemRuntimeGetScriptContainer) // push the script container (executing transaction, actually).
+	emit.Opcodes(vrf.BinWriter, opcode.PUSH0, opcode.PICKITEM)                // pick 0-th transaction item (the transaction hash).
+	// Concatenate network magic and transaction hash.
+	emit.Opcodes(vrf.BinWriter, opcode.CAT)    // this instruction will convert network magic to bytes using BigInteger rules of conversion.
+	emit.Opcodes(vrf.BinWriter, opcode.STLOC2) // store msg as a local variable #2.
+
+	// Initialize outer local variables: numOk, pubsStart, i.
+	emit.Opcodes(vrf.BinWriter, opcode.PUSH0, opcode.STLOC3, // initialize numOk.
+		opcode.PUSH0, opcode.STLOC4, // initialize pubsStart.
+		opcode.PUSH0, opcode.STLOC5) // initialise i.
+
+	// Start sigs (i) iteration.
+	iIterationStartOffset := vrf.Len()
+	emit.Opcodes(vrf.BinWriter, opcode.LDLOC5)               // load i.
+	emit.Int(vrf.BinWriter, int64(m))                        // push m.
+	emit.Instruction(vrf.BinWriter, opcode.JMPGE, []byte{0}) // jump to the end of the script if i >= m.
+	iIterationEndOffset := vrf.Len()
+	emit.Opcodes(vrf.BinWriter, opcode.LDLOC4, opcode.STLOC6) // initialize inner local variable j = pubsStart.
+
+	// Start pubs (j) iteration.
+	jIterationStartOffset := vrf.Len()
+	emit.Opcodes(vrf.BinWriter, opcode.LDLOC6, // load j.
+		opcode.LDLOC1, opcode.SIZE) // load pubs and get its size. TODO: may be optimized by storing size as LOC7, but it requires LOC+param, need to compare prices.
+	emit.Instruction(vrf.BinWriter, opcode.JMPGE, []byte{0}) // jump to the next (i+1)-th iteration over sigs if j >= len(pubs).
+	jIterationEndOffset := vrf.Len()
+
+	// Main loop inner code goes below.
+
+	// Prepare arguments and call CryptoLib's verifyWithECDsa.
+	emit.Int(vrf.BinWriter, int64(native.Secp256k1Keccak256)) // push Koblitz curve identifier.
+	emit.Opcodes(vrf.BinWriter, opcode.LDLOC0,                // load signatures.
+		opcode.LDLOC5,             // load i.
+		opcode.PICKITEM,           // pick signature at index i.
```

Comment:
Maybe you can save on some counters if you're to POPITEM here...

Reply:
Or maybe it'll complicate code for no real benefit (1% GAS cost improvement isn't worth it).

Reply:
For pubs doesn't worth it: [popitem_for_pubs_patch.txt](https://github.com/nspcc-dev/neo-go/files/15209498/popitem_for_pubs_patch.txt)
For sigs cost will be larger since some additional management is required to decide is we're to keep or pop the sig.